tket2/src/circuit.rs

```diff
@@ -147,14 +147,31 @@ impl<T: HugrView> Circuit<T> {
             .expect("Circuit has no I/O nodes")
     }
 
-    /// The number of quantum gates in the circuit.
+    /// The number of operations in the circuit.
+    ///
+    /// This includes [`Tk2Op`]s, pytket ops, and any other custom operations.
+    ///
+    /// Nested circuits are traversed to count their operations.
+    ///
+    ///   [`Tk2Op`]: crate::Tk2Op
     #[inline]
-    pub fn num_gates(&self) -> usize
+    pub fn num_operations(&self) -> usize
     where
         Self: Sized,
     {
-        // TODO: Discern quantum gates in the commands iterator.
-        self.hugr().children(self.parent).count() - 2
+        let mut count = 0;
+        let mut roots = vec![self.parent];
+        while let Some(node) = roots.pop() {
```

Comment:
There's `hugr::DescendantsGraph` but that's optimised for arbitrary accesses to the HugrView.
It keeps a cache of nodes it knows are in graph in a `RefCell<HashMap<_,_>>`, and updates it as it filters throught.
We can skip that here since we have a set traversal order.

I guess we could optimise that in Portgraph. I'll open an issue, but leave the explicit loop here for now.

Reply:
why not just iterate over all nodes in the hugr?

Reply:
Leaving this here to register the backlink:
https://github.com/CQCL/portgraph/issues/135